File: 3225. maximum-score-from-grid-operations.java
import java.util.*;

class Solution {
    public long maximumScore(int[][] A) {
        int n = A.length;

        long[][] B = new long[n][n + 1];
        for (int j = 0; j < n; j++) {
            for (int i = 0; i < n; i++) {
                B[j][i + 1] = B[j][i] + A[i][j];
            } 
        }

        long[] pdp = new long[n + 1];
        long[] pep = new long[n + 1];

        for (int j = 1; j < n; j++) {
            long[] dp = new long[n + 1];
            long[] ep = new long[n + 1];

            for (int pb = 0; pb <= n; pb++) {
                for (int cb = 0; cb <= n; cb++) {

                    long pv = 0;
                    if (cb > pb) {
                        pv = B[j - 1][cb] - B[j - 1][pb];
                    }

                    long cv = 0;
                    if (cb < pb) {
                        cv = B[j][pb] - B[j][cb];
                    }

                    dp[cb] = Math.max(dp[cb],
                            Math.max(pv + pdp[pb], pep[pb])
                    );

                    ep[cb] = Math.max(ep[cb],
                            Math.max(cv + pep[pb],
                                     cv + pv + pdp[pb])
                    );
                }
            }

            pdp = dp;
            pep = ep;
        }

        long ans = 0;
        for (long v : pep) {
            ans = Math.max(ans, v);
        }

        return ans;
    }
    
    // Test cases
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Example 1
        int[][] grid1 = {{0,0,0,0,0},{0,0,3,0,0},{0,1,0,0,0},{5,0,0,3,0},{0,0,0,0,2}};
        long result1 = solution.maximumScore(grid1);
        System.out.println("Example 1: Input: grid = [[0,0,0,0,0],[0,0,3,0,0],[0,1,0,0,0],[5,0,0,3,0],[0,0,0,0,2]]");
        System.out.println("Output: " + result1);
        System.out.println("Expected: 11\n");
        
        // Example 2
        int[][] grid2 = {{10,9,0,0,15},{7,1,0,8,0},{5,20,0,11,0},{0,0,0,1,2},{8,12,1,10,3}};
        long result2 = solution.maximumScore(grid2);
        System.out.println("Example 2: Input: grid = [[10,9,0,0,15],[7,1,0,8,0],[5,20,0,11,0],[0,0,0,1,2],[8,12,1,10,3]]");
        System.out.println("Output: " + result2);
        System.out.println("Expected: 94");
    }
}

File: 3225. maximum-score-from-grid-operations.py
class Solution:
    def maximumScore(self, grid: List[List[int]]) -> int:
        

        prefix = [[0] * (len(grid) + 1)]

        for i in range(len(grid) + 1):
            prefix.append([0] * (len(grid) + 1))
        
        # print(prefix)

        for i in range(len(grid)):
            sm = 0
            for j in range(len(grid[0])):
                sm += grid[j][i]
                prefix[j + 1][i] = sm

        # print(prefix)

        @cache
        def dp(col, prev, flag):

            if col >= len(grid):
                return 0

            
            ans = 0 
            
            for i in range(0, len(grid) + 1):


                if prev > i:
                    sm = 0
                    sm = (prefix[prev][col] - prefix[i][col])
                    ans = max(ans, sm + dp(col + 1, i, True))
                elif i > prev and not flag and col > 0:
                    sm_prev = 0
                    # for r in range(prev, i):
                    # print(i, prev)
                    # print(prefix[i][col], prefix[prev][col])
                    sm_prev = (prefix[i][col - 1] - prefix[prev][col - 1])
                    ans = max(ans, sm_prev + dp(col + 1, i, False))
                else:
                    ans = max(ans, 0 + dp(col + 1, i, False))

            
            return ans 
        

        tmp = dp(0, 0, True)
        dp.cache_clear()
        return (tmp)


# Test cases
if __name__ == "__main__":
    from typing import List
    from functools import cache
    
    solution = Solution()
    
    # Example 1
    grid1 = [[0,0,0,0,0],[0,0,3,0,0],[0,1,0,0,0],[5,0,0,3,0],[0,0,0,0,2]]
    result1 = solution.maximumScore(grid1)
    print(f"Example 1: Input: grid = [[0,0,0,0,0],[0,0,3,0,0],[0,1,0,0,0],[5,0,0,3,0],[0,0,0,0,2]]")
    print(f"Output: {result1}")
    print(f"Expected: 11\n")
    
    # Example 2
    grid2 = [[10,9,0,0,15],[7,1,0,8,0],[5,20,0,11,0],[0,0,0,1,2],[8,12,1,10,3]]
    result2 = solution.maximumScore(grid2)
    print(f"Example 2: Input: grid = [[10,9,0,0,15],[7,1,0,8,0],[5,20,0,11,0],[0,0,0,1,2],[8,12,1,10,3]]")
    print(f"Output: {result2}")
    print(f"Expected: 94")